src/exception_summary.jl

```diff
@@ -0,0 +1,98 @@
+
+Print a summary of the current task's exceptions to `io`.
+
+This is particularly helpful in cases where the exception stack is large, the backtraces are
+large, and CompositeExceptions with multiple parts are involved.
+"""
+function summarize_current_exceptions(io::IO=Base.stderr)
+    printstyled(io, TITLE, '\n'; color=Base.info_color())
+    _summarize_task_exceptions(io, current_task(), 0)
+    return nothing
+end
+
+function _indent_print(io::IO, indent::Int, x...; color=:normal)
+    printstyled(io, " "^indent, x...; color=color)
+end
+
+function _indent_println(io::IO, indent::Int, x...; color=:normal)
+    _indent_print(io, indent, x..., "\n"; color=color)
+end
+
+function _summarize_task_exceptions(io::IO, task::Task, indent::Int)
+    exception_stack = current_exceptions(task)
+    for (i, es) in enumerate(exception_stack)
+        e, stack = es
+        if i != 1
+            # TODO: should the indention increase here?
+            println(io)
+            _indent_println(io, indent, "which caused:"; color=Base.error_color())
+        end
+        _summarize_exception(io, e, stack, indent)
+    end
+end
+
+"""
+    _summarize_exception(io::IO, e::TaskFailedException, _, indent::Int)
+    _summarize_exception(io::IO, e::CompositeException, stack, indent::Int)
+    _summarize_exception(io::IO, e::Exception, stack, indent::Int)
+
+The secret sauce that lets us unwrap TaskFailedExceptions and CompositeExceptions, and
+summarize the actual exception.
+
+TaskFailedException simply wraps a task, so it is just unwrapped, and processed by
+_summarize_task_exceptions().
+
+CompositeException simply wraps a Vector of Exceptions. Each of the individual Exceptions is
+summarized.
+
+All other exceptions are printed via [`Base.showerror()`](@ref). The first stackframe in the
+backtrace is also printed.
+"""
+function _summarize_exception(io::IO, e::TaskFailedException, _, indent::Int)
+    # recurse down the exception stack to find the original exception
+    _summarize_task_exceptions(io, e.task, indent)
+end
+function _summarize_exception(io::IO, e::CompositeException, stack, indent::Int)
+    _indent_println(io, indent, "CompositeException (length ", length(e), "):")
+    for (i, ex) in enumerate(e.exceptions)
+        _summarize_exception(io, ex, stack, indent + INDENT_LENGTH)
+        # print something to separate the multiple exceptions wrapped by CompositeException
+        if i != length(e.exceptions)
+            _indent_println(io, indent + INDENT_LENGTH, SEPARATOR)
+        end
+    end
+end
+# This is the overload that prints the actual exception that occurred.
+function _summarize_exception(io::IO, exc, stack, indent::Int)
+    # Print the exception.
+    _indent_print(io, indent)
+    Base.showerror(io, exc)
+    println(io)
+
+    # Print the source line number of the where the exception occurred.
+    bt = Base.process_backtrace(stack)
+    # borrowed from julia/base/errorshow.jl
+    (frame, n) = bt[1]
```

Comment:
In the compilation profiling project, we discovered that just calling `Base.process_backtrace()` can be really slow in prod for us, since our binary is so big!

So i think it's better to grab just the first frame from the backtrace _first_, and then either leave all the inlined names, or still filter it one more time:
```suggestion
    bt = Base.process_backtrace(stack[1:1])
    # borrowed from julia/base/errorshow.jl
    (frame, n) = bt[1]
```

Reply:
🤔 but i also seem to remember that we tried something like this but it didn't always work.. 🤔 lemme play with it locally

Reply:
Hokay, i think i got something that should work, here: e1fa423

Reply:
👍 that commit LGTM. Thanks!